xds/src/main/java/io/grpc/xds/XdsClient.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import io.grpc.Status;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * An {@link XdsClient} instance encapsulates all of the logic for communicating with the xDS
+ * server. It may create multiple RPC streams for a series of xDS protocols (e.g., LDS, RDS,
+ * VHDS, CDS and EDS) over a single channel. Watch-based interfaces are provided for each set
+ * of data needed by gRPC.
+ *
+ * <p>This class should only be instantiated by the xDS resolver but can be passed to load
+ * balancing policies.
+ */
+abstract class XdsClient {
+
+  /**
+   * Data class containing the results of performing a series of resource discovery RPCs via
+   * LDS/RDS/VHDS protocols. The results may include configurations for path/host rewriting,
+   * traffic mirroring, retry or hedging, default timeouts and load balancing policy that will
+   * be used to generate a service config.
+   */
+  // TODO(chengyuanzhang): content TBD, most information comes from VirtualHost proto.
+  static final class ConfigUpdate {
+
+  }
+
+  // TODO(zdapeng): content TBD.
+  static final class ClusterUpdate {
+
+  }
+
+  // TODO(zdapeng): content TBD.
+  static final class EndpointUpdate {
+
+  }
+
+  /**
+   * Config watcher interface. To be implemented by the xDS resolver.
+   */
+  @NotThreadSafe
+  interface ConfigWatcher {
+
+    /**
+     * Called when receiving an update on virtual host configurations.
+     */
+    void onConfigChanged(ConfigUpdate update);
+
+    void onError(Status error);
```

Comment:
`onError(Throwable)`. The major usecase should be on exceptions, this is the case even for `StreamObserver.onError()`

Reply:
`Status` can carry more information than `Throwable`, which would be useful for objects watching the data. It's not a big issue, we can change this any time if we found incompatibilities.

Reply:
You will have to make up a Status code when you catch a exception and then call `onError(Status)`, and this "more information" Status code is useless for ``onError()` to handle.

Reply:
You can always convert `Throwable` to `Status` by `Status.fromThrowable(...)`. The status code is `UNKNOWN` but `Status` has some useful methods that can be used to augment error message.

Reply:
That's fare.